units/en/unit2/smolagents/code_agents.mdx

```diff
@@ -0,0 +1,292 @@
+
+A `CodeAgent` performs actions through a cycle of steps, with existing variables and knowledge being incorporated into the agent's logs as follows:
+
+1. The system prompt is stored in a `SystemPromptStep`, and the user query is logged in a `TaskStep`.
+
+2. Then, the following while loop is executed:
+
+    2.1 The method `agent.write_memory_to_messages()` writes the agent's logs into a list of LLM-readable [chat messages](https://huggingface.co/docs/transformers/en/chat_templating).
+    
+    2.2 These messages are sent to a `Model`, which generates a completion. 
+    
+    2.3 The completion is parsed to extract the action, which, in our case, could be a code snippet since we're working with a `CodeAgent`.
+    
+    2.4 The action is executed.
+    
+    2.5 The results are logged into memory in an `ActionStep`.
+
+At the end of each step, if the agent includes any function calls (in `agent.step_callback`), they are executed. Below, you can see an comparison diagram between a multi-step agent using the ReAct framework and a one step agent.
+
+![Comparison diagram between a multi-step agent using the ReAct framework and a one step agent](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/open-source-llms-as-agents/ReAct.png)
+
+## Let's See Some Examples
+
+Alfred is planning a party at the Wayne family mansion and needs your help to ensure everything goes smoothly. To assist him, we'll apply what we've learned about how a multi-step `CodeAgent` operates.
+
+If you haven't installed `smolagents` yet, you can do so by running the following command:
+
+```bash
+pip install smolagents -U
+```
+
+Let's also login to the Hugging Face Hub to have access to the Inference API.
+
+```python
+from huggingface_hub import login
+
+login()
+```
+
+### Selecting a Playlist for the Party Using `smolagents`
+
+An important part of a successful party is the music. Alfred needs some help selecting the playlist. Luckily, `smolagents` has got us covered! We can build an agent capable of searching the web using DuckDuckGo. To give the agent access to this tool, we include it in the tool list when creating the agent.
+
+For the model, we'll rely on `HfApiModel`, which provides access to Hugging Face's [Inference API](https://huggingface.co/docs/api-inference/index). The default model is `"Qwen/Qwen2.5-Coder-32B-Instruct"`, which is performant and available for fast inference, but you can select any compatible model from the Hub.
+
+Running an agent is quite straightforward:
+
+```python
+from smolagents import CodeAgent, DuckDuckGoSearchTool, HfApiModel
+
+agent = CodeAgent(tools=[DuckDuckGoSearchTool()], model=HfApiModel())
+
+agent.run("Search for the best music recommendations for a party at the Wayne's mansion.")
+```
+
+When you run this example, the output will display a trace of the workflow steps being executed. It will also print the corresponding Python code with the message: 
```

Comment:
```suggestion
When you run this example, the output will **display a trace of the workflow steps being executed**. It will also print the corresponding Python code with the message: 
```